File: services/paper_ingestion/paper_ingestion/models/papers.py
"""Paper-centric Pydantic models.

Core paper records, chunks, summaries, user state, feed, feedback,
priority, discovery, search, and paper-source configuration.
"""

from datetime import date, datetime
from enum import Enum
from typing import Any, Literal

from jarvis_common.crypto import mask_secret
from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

# Keys whose values must never appear in plaintext in API responses.
_SECRET_KEY_NAMES: frozenset[str] = frozenset(
    {"api_key", "client_secret", "token", "password", "secret", "bearer"}
)

# --- Enums ---


class SourceType(str, Enum):
    """Supported paper source types."""

    ARXIV = "arxiv"
    SEMANTIC_SCHOLAR = "semantic_scholar"
    LOCAL = "local"
    OPENALEX = "openalex"
    PUBMED = "pubmed"
    ZOTERO = "zotero"


class Confidence(str, Enum):
    """Summary confidence level based on quote verification pass rate.

    NONE = no findings to verify, HIGH = 100% verified, MEDIUM = >50%, LOW = <=50%.
    """

    NONE = "NONE"
    HIGH = "HIGH"
    MEDIUM = "MEDIUM"
    LOW = "LOW"


# --- Core Paper Models ---


class PaperBase(BaseModel):
    """Fields common to paper creation and response."""

    external_id: str = Field(..., max_length=255)
    source_type: SourceType
    title: str = Field(..., max_length=1000)
    authors: list[str]
    abstract: str | None = Field(default=None, max_length=50000)
    published_date: date | None = None
    url: str = Field(..., max_length=2000)
    pdf_url: str | None = Field(default=None, max_length=2000)
    citation_count: int = 0
    metadata: dict = Field(default_factory=dict)

    @field_validator("url", "pdf_url")
    @classmethod
    def validate_url(cls, v: str | None) -> str | None:
        if v is None:
            return v
        if not v.startswith(("http://", "https://", "local://")):
            raise ValueError("URL must start with http://, https://, or local://")
        return v


class PaperCreate(PaperBase):
    """Used when inserting a new paper from a source."""

    # H.11 cross-ref: the literal union below is mirrored in
    # frontend/src/types/index.ts ('inbox'|'to_read'|'reading'|'done'|'trash'
    # for LifecycleState plus the discovery-origin enum below). When you
    # add or rename a discovery_origin value here, also update the
    # frontend type and any constants in frontend/src/components/feed/.
    discovery_origin: Literal["user_initiated", "pulse", "recommender", "citation_batch"] = (
        "user_initiated"
    )


class PaperResponse(PaperBase):
    """Full paper representation returned by the API."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    discovery_origin: Literal["user_initiated", "pulse", "recommender", "citation_batch"] = (
        "user_initiated"  # immutable after insert
    )
    pdf_local_path: str | None = None
    pdf_downloaded: bool = False
    discovered_at: datetime | None = None
    priority_score: float | None = None
    created_at: datetime


class ChunkResponse(BaseModel):
    """A single text chunk from a processed paper."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    paper_id: int
    chunk_index: int
    content: str
    page_number: int | None = None
    start_char: int | None = None
    end_char: int | None = None
    embedding_id: str | None = None
    created_at: datetime


class KeyFinding(BaseModel):
    """A single finding within a paper summary.

    The ``verified`` flag starts as False and is set by the quote
    verification pipeline in ``verification.py``.
    """

    finding: str
    quote: str
    page_number: int | None = None
    chunk_id: int | None = None
    verified: bool = False
    snapshot_path: str | None = None


class CrossReference(BaseModel):
    """A link between two related papers discovered via cross-reference check."""

    related_paper_id: int
    relationship: str
    explanation: str
    related_quote: str | None = None
    content_generation: int = 0


class SummaryResponse(BaseModel):
    """LLM-generated summary with verified citations."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    paper_id: int
    summary_brief: str
    summary_detailed: str
    tldr: str | None = None
    key_findings: list[KeyFinding]
    methodology: str | None = None
    limitations: str | None = None
    relevance_notes: str | None = None
    confidence: Confidence
    cross_references: list[CrossReference]
    llm_model: str | None = None
    summary_verified: bool = False
    created_at: datetime
    # Not persisted — set only on fresh summarize responses.
    coverage: float | None = None
    passes: int | None = None


class UserStateResponse(BaseModel):
    """User reading state for a paper.

    H.11 cross-ref: the ``state`` and ``state_before_trash`` literal unions
    below MUST stay in sync with ``LifecycleState`` and ``StateBeforeTrash``
    in ``frontend/src/types/index.ts``. The frontend uses the same string
    values verbatim, so any addition / rename / removal here is a
    breaking schema change requiring a coordinated frontend update.
    """

    state: Literal["inbox", "to_read", "reading", "done", "trash"]
    state_before_trash: Literal["inbox", "to_read", "reading", "done"] | None = None
    starred: bool = False
    rating: int | None = None
    user_notes: str | None = None
    flagged: bool = False
    updated_at: datetime | None = None


class RecentFeedback(BaseModel):
    """Most recent feedback by current user (UI affordance state)."""

    signal: Literal["positive", "negative"]
    source: str
    created_at: datetime


class PaperDetailResponse(BaseModel):
    """Paper with its summary, chunks, user state, and most recent feedback."""

    paper: PaperResponse
    summary: SummaryResponse | None = None
    chunks: list[ChunkResponse] = Field(default_factory=list)
    user_state: UserStateResponse | None = None
    recent_feedback: RecentFeedback | None = None
    has_project_links: bool = False
    # True when the most recent paper.process / paper.analyze job for this
    # paper+user terminated in `failed` (procrastinate_jobs.status). Lets the
    # left Pipeline rail (PaperTOC) show ✗ from the SAME persisted failure
    # source ActionsSidebar already polls via getJob — no parallel status.
    processing_failed: bool = False


# --- Search / Request Models ---


class SearchRequest(BaseModel):
    """Request body for POST /api/search and POST /api/search-preview.

    Supports both legacy single-source format and new multi-source format:
    - Legacy: ``{"source": "arxiv", ...}``
    - Legacy alias: ``{"source": "both", ...}`` → expands to arxiv + semantic_scholar
    - New: ``{"source_types": ["arxiv", "pubmed"], ...}``
    """

    query: str = Field(..., min_length=1, max_length=500)
    # Legacy field kept for backward compat; migrated to source_types by validator.
    source: SourceType | None = None
    source_types: list[SourceType] = Field(
        default_factory=lambda: [SourceType.ARXIV],
        min_length=1,
    )
    max_results: int = Field(default=10, ge=1, le=200)
    year_from: int | None = Field(default=None, ge=1900, le=2100)
    year_to: int | None = Field(default=None, ge=1900, le=2100)
    sort_by: Literal["relevance", "date"] = "relevance"
    author: str | None = Field(default=None, max_length=200)

    @model_validator(mode="before")
    @classmethod
    def _migrate_legacy_source(cls, data: Any) -> Any:
        """Migrate legacy ``source`` field to ``source_types``.

        Handles:
        - ``{"source": "both"}`` → ``{"source_types": ["arxiv", "semantic_scholar"]}``
        - ``{"source": "arxiv"}`` → ``{"source_types": ["arxiv"]}``
        - ``{"source_types": [...]}`` → pass through unchanged
        """
        if not isinstance(data, dict):
            return data
        if "source_types" not in data and "source" in data:
            source_val = data.get("source")
            if source_val == "both":
                data = dict(data)
                data["source_types"] = ["arxiv", "semantic_scholar"]
                data.pop("source", None)
            elif source_val is not None:
                data = dict(data)
                data["source_types"] = [source_val]
                # Keep source for backward compat (callers may still read it)
        return data


# --- Internal Models ---


class ChunkForEmbedding(BaseModel):
    """A chunk prepared for embedding (before DB insertion)."""

    chunk_index: int
    content: str
    page_number: int | None = None
    start_char: int
    end_char: int


class PaperSourceConfig(BaseModel):
    """Configuration row from the paper_sources table."""

    id: int
    source_type: SourceType
    enabled: bool
    config: dict = Field(default_factory=dict)


class SourceResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    source_type: SourceType
    enabled: bool
    config: dict = Field(default_factory=dict)
    priority: int = 1
    display_order: int = 0
    created_at: datetime

    @model_validator(mode="after")
    def _redact_secrets(self) -> "SourceResponse":
        """Mask known secret keys in config so they are never leaked in API responses."""
        if isinstance(self.config, dict):
            redacted = {
                k: (mask_secret(str(v)) if k.lower() in _SECRET_KEY_NAMES and v else v)
                for k, v in self.config.items()
            }
            object.__setattr__(self, "config", redacted)
        return self


class SourceUpdate(BaseModel):
    enabled: bool | None = None
    config: dict | None = None
    priority: int | None = None
    display_order: int | None = None


# --- Feed Models ---


class FeedPaper(PaperResponse):
    """Paper with joined summary and user-state fields for the feed.

    H.11 cross-ref: ``state`` / ``state_before_trash`` mirror the
    ``LifecycleState`` / ``StateBeforeTrash`` types in
    ``frontend/src/types/index.ts``. Keep them in lockstep.
    """

    summary_brief: str | None = None
    tldr: str | None = None
    confidence: Confidence | None = None
    state: Literal["inbox", "to_read", "reading", "done", "trash"] = "inbox"
    state_before_trash: Literal["inbox", "to_read", "reading", "done"] | None = None
    starred: bool = False
    rating: int | None = None
    priority_level: str | None = None
    has_chunks: bool = False
    has_summary: bool = False
    recommendation_score: float | None = None
    recommendation_reason: str | None = None
    recommendation_modes: list[str] | None = None
    note_match_count: int = 0
    note_snippet: str | None = None
    recent_feedback: RecentFeedback | None = None


class FeedResponse(BaseModel):
    """Paginated response for the What's New paper feed."""

    papers: list[FeedPaper]
    total: int
    search_mode: str = "filtered"


class DiscoverRequest(BaseModel):
    """Request body for seed-based paper discovery."""

    paper_ids: list[int] = Field(..., min_length=1, max_length=10)
    limit: int = Field(default=10, ge=1, le=50)
    score_threshold: float = Field(default=0.5, ge=0.0, le=1.0)


# --- Priority Helpers ---


def compute_priority(
    relevance_scores: list[float],
    discovered_at: datetime | None,
    citation_count: int | None,
    now: datetime,
) -> float:
    """Compute composite priority score for a paper.

    Parameters
    ----------
    relevance_scores : list[float]
        Relevance scores from all linked topics.
    discovered_at : datetime or None
        When the paper was discovered.
    citation_count : int or None
        Number of citations.
    now : datetime
        Current time for recency calculation.

    Returns
    -------
    float
        Priority score between 0.0 and 1.0.
    """
    relevance = max(relevance_scores) if relevance_scores else 0.0
    recency = 0.5  # Default if no discovered_at
    if discovered_at:
        days_old = (now - discovered_at).total_seconds() / 86400
        recency = max(0.0, 1.0 - days_old / 30)
    citation_boost = min(1.0, (citation_count or 0) / 100)
    return round(0.5 * relevance + 0.3 * recency + 0.2 * citation_boost, 4)


def priority_level(score: float | None) -> str:
    """Convert priority score to display level.

    Parameters
    ----------
    score : float or None
        Priority score between 0.0 and 1.0.

    Returns
    -------
    str
        One of ``"must-read"``, ``"recommended"``, ``"background"``, or ``"unscored"``.
    """
    if score is None:
        return "unscored"
    if score > 0.7:
        return "must-read"
    if score > 0.4:
        return "recommended"
    return "background"


# --- Endpoint Response Models ---


class SystemModelsResponse(BaseModel):
    """Response for GET /api/system/models."""

    status: str
    installed: list[dict[str, Any]]
    hardware: dict[str, Any]
    current: dict[str, Any]
    issues: dict[str, str]
    catalog: list[dict[str, Any]] = Field(default_factory=list)
    recommendations: dict[str, list[dict[str, Any]]] = Field(default_factory=dict)
    reviewed_choices: dict[str, list[dict[str, Any]]] = Field(default_factory=dict)
    embedding_contract: dict[str, Any] = Field(default_factory=dict)
    # Advisory per-VRAM default-model recommendation.  Always present; never
    # mutates config.  confirm_on_target=True on individual aliases means the
    # recommendation has not been validated via a live bench on the target GPU.
    hardware_recommendation: dict[str, Any] = Field(default_factory=dict)


class ProcessPdfResponse(BaseModel):
    """Response for POST /api/process-pdf/{paper_id}."""

    paper_id: int
    chunk_count: int
    status: str


class ScanLocalPdfsResponse(BaseModel):
    """Response for POST /api/scan-local-pdfs."""

    scanned: int
    imported: int
    skipped: int


class MarkReadResponse(BaseModel):
    """Response for PUT /api/papers/{paper_id}/read."""

    status: str
    paper_id: int


class RelevanceScoreResponse(BaseModel):
    """Response for POST /api/relevance-score."""

    paper_id: int
    topic_id: int
    relevance_score: float


class FeedbackRequest(BaseModel):
    """Body for POST /api/papers/{paper_id}/feedback."""

    signal: Literal["positive", "negative"]
    source: Literal[
        "pulse_thumbs",
        "feed_thumbs",
        "paper_detail_thumbs",
        "dismiss_combined",
    ]
    reason: str | None = None


class FeedbackResponse(BaseModel):
    """Response for POST /api/papers/{paper_id}/feedback."""

    paper_id: int
    signal: Literal["positive", "negative"]
    source: str
    created_at: datetime


class PaperPriorityResponse(BaseModel):
    """Response for POST /api/papers/{paper_id}/priority."""

    paper_id: int
    priority_score: float
    priority_level: str


class RecomputePrioritiesResponse(BaseModel):
    """Response for POST /api/papers/recompute-priorities."""

    updated: int


class FeedbackListItem(BaseModel):
    """One row in GET /api/recommendation_feedback response."""

    paper_id: int
    title: str
    signal: Literal["positive", "negative"]
    source: Literal[
        "pulse_thumbs",
        "feed_thumbs",
        "paper_detail_thumbs",
        "dismiss_combined",
    ]
    reason: str | None = None
    topic_id: int | None = None
    topic_name: str | None = None
    created_at: datetime


class FeedbackListResponse(BaseModel):
    """Response for GET /api/recommendation_feedback."""

    items: list[FeedbackListItem]
    total: int


class DeleteFeedbackResponse(BaseModel):
    """Response for DELETE /api/recommendation_feedback/{topic_id}."""

    deleted: int
    topic_id: int


class BatchProcessResponse(BaseModel):
    """Response for POST /api/papers/batch-process."""

    queued: int
    total_unprocessed: int
    skipped_missing_pdf: int
    job_id: str | None = None


class BatchSummarizeResponse(BaseModel):
    """Response for POST /api/papers/batch-summarize.

    ``job_id`` is ``None`` when no unsummarised papers were found.
    """

    total_unsummarized: int
    job_id: str | None = None


class WeeklyDigestResponse(BaseModel):
    """Response for GET /api/digest/weekly."""

    topics: list[dict[str, Any]] = Field(default_factory=list)
    total_papers: int
    period_start: str
    period_end: str


class DiscoveryResultItem(BaseModel):
    """A single discovered paper in POST /api/discover response."""

    paper_id: int
    title: str
    authors: list[str]
    url: str
    similarity_score: float
    matching_snippet: str = ""


class PaperBriefResponse(BaseModel):
    """A lightweight paper entry for selector dropdowns."""

    id: int
    title: str
    source_type: str | None = None
    published_date: date | None = None


# --- Hybrid Search / Similar Papers Models ---


class HybridSearchResult(BaseModel):
    """A single result from POST /api/papers/search-hybrid."""

    id: int
    title: str
    authors: list[str]
    url: str
    abstract: str | None = None
    published_date: date | None = None
    rrf_score: float
    bm25_rank: int | None = None
    semantic_rank: int | None = None


class SimilarPaperResult(BaseModel):
    """A single result from GET /api/similar/{paper_id}."""

    paper_id: int
    title: str
    authors: list[str]
    url: str
    similarity_score: float
    matching_snippet: str = ""


# --- Analytics Models ---


class PapersBySourceItem(BaseModel):
    """A single row from GET /analytics/papers-by-source."""

    source_type: str
    count: int


class PapersByStatusItem(BaseModel):
    """A single row from GET /analytics/papers-by-status."""

    status: str
    count: int


# --- Paper Lifecycle Action Models ---


class BulkActionRequest(BaseModel):
    """Body for POST /api/papers/bulk."""

    paper_ids: list[int] = Field(..., min_length=1, max_length=500)
    action: Literal[
        "save",
        "skip",
        "trash",
        "mark_reading",
        "mark_done",
        "restore",
        "star",
        "unstar",
        "feedback_positive",
        "feedback_negative",
        "hard_delete",  # bulk permanent delete (only valid on state='trash')
    ]


class BulkActionFailure(BaseModel):
    """A single failed paper in a POST /api/papers/bulk response."""

    paper_id: int
    error: str


class BulkActionResponse(BaseModel):
    """Response for POST /api/papers/bulk.

    Partial failures are reported per-paper; ``failed`` carries a safe,
    operator-diagnostic ``error`` code (never raw exception text).
    """

    succeeded: list[int] = Field(default_factory=list)
    failed: list[BulkActionFailure] = Field(default_factory=list)


class TopicFacetCount(BaseModel):
    """Per-topic count for the feed facet rail (§ Topic section)."""

    topic_id: int
    name: str
    count: int


class FeedCountsResponse(BaseModel):
    """Response for GET /api/papers/feed/counts (10 named views).

    UI v3 additive facets (by_source, by_topic, untagged) are scoped to the
    caller's user_library exactly as the named-view counts above.
    """

    inbox: int
    library: int
    reading_list: int
    reading: int
    done: int
    starred: int
    trash: int
    active: int
    kept: int
    all_non_trash: int
    # UI v3 facet rail — additive, always present (empty when no library).
    by_source: dict[str, int] = Field(default_factory=dict)
    by_topic: list[TopicFacetCount] = Field(default_factory=list)
    untagged: int = 0


class AnnotationsRequest(BaseModel):
    """Body for PUT /api/papers/{paper_id}/annotations."""

    rating: int | None = Field(default=None, ge=1, le=5)
    user_notes: str | None = None
    flagged: bool | None = None


class ProcessBatchRequest(BaseModel):
    """Body for POST /api/papers/process_batch."""

    paper_ids: list[int] = Field(..., min_length=1, max_length=50)

File: frontend/src/lib/api/schemas/papers.ts
import { z } from 'zod';
import { jsonObjectSchema, jsonValueSchema } from './common';

export const sourceTypeSchema = z.enum([
  'arxiv',
  'semantic_scholar',
  'local',
  'openalex',
  'pubmed',
  'zotero',
]);

export const confidenceSchema = z.enum(['NONE', 'HIGH', 'MEDIUM', 'LOW']);
export const discoveryOriginSchema = z.enum([
  'user_initiated',
  'pulse',
  'recommender',
  'citation_batch',
]);
export const lifecycleStateSchema = z.enum(['inbox', 'to_read', 'reading', 'done', 'trash']);
export const stateBeforeTrashSchema = z.enum(['inbox', 'to_read', 'reading', 'done']).nullable();
export const feedbackSourceSchema = z.enum([
  'pulse_thumbs',
  'feed_thumbs',
  'paper_detail_thumbs',
  'dismiss_combined',
]);

export const recentFeedbackSchema = z.looseObject({
  signal: z.enum(['positive', 'negative']),
  source: feedbackSourceSchema,
  created_at: z.string(),
});

export const paperSchema = z.looseObject({
  id: z.number(),
  external_id: z.string(),
  source_type: sourceTypeSchema,
  title: z.string(),
  authors: z.array(z.string()),
  abstract: z.string().nullable(),
  published_date: z.string().nullable(),
  url: z.string(),
  pdf_url: z.string().nullable(),
  pdf_local_path: z.string().nullable(),
  pdf_downloaded: z.boolean(),
  citation_count: z.number(),
  priority_score: z.number().nullable(),
  metadata: jsonObjectSchema,
  discovered_at: z.string().nullable(),
  created_at: z.string(),
  discovery_origin: discoveryOriginSchema,
  recent_feedback: recentFeedbackSchema.nullable().optional(),
});

export const userStateSchema = z.looseObject({
  state: lifecycleStateSchema,
  state_before_trash: stateBeforeTrashSchema,
  starred: z.boolean(),
  rating: z.number().nullable(),
  user_notes: z.string().nullable(),
  flagged: z.boolean(),
  updated_at: z.string().nullable(),
});

const keyFindingSchema = z.looseObject({
  finding: z.string(),
  quote: z.string(),
  page_number: z.number().nullable(),
  chunk_id: z.number().nullable(),
  verified: z.boolean(),
  snapshot_path: z.string().nullable(),
});

const crossReferenceSchema = z.looseObject({
  related_paper_id: z.number(),
  relationship: z.string(),
  explanation: z.string(),
  related_quote: z.string().nullable(),
  content_generation: z.number().optional(),
});

const summarySchema = z.looseObject({
  id: z.number(),
  paper_id: z.number(),
  summary_brief: z.string(),
  summary_detailed: z.string(),
  tldr: z.string().nullable(),
  key_findings: z.array(keyFindingSchema),
  methodology: z.string().nullable(),
  limitations: z.string().nullable(),
  relevance_notes: z.string().nullable(),
  confidence: confidenceSchema,
  cross_references: z.array(crossReferenceSchema),
  llm_model: z.string().nullable(),
  summary_verified: z.boolean(),
  created_at: z.string(),
  coverage: z.number().nullable().optional(),
  passes: z.number().nullable().optional(),
});

const chunkSchema = z.looseObject({
  id: z.number(),
  paper_id: z.number(),
  chunk_index: z.number(),
  content: z.string(),
  page_number: z.number().nullable(),
  start_char: z.number().nullable(),
  end_char: z.number().nullable(),
  embedding_id: z.string().nullable(),
  created_at: z.string(),
});

export const paperDetailSchema = z.looseObject({
  paper: paperSchema,
  summary: summarySchema.nullable(),
  chunks: z.array(chunkSchema),
  user_state: userStateSchema.nullable(),
  recent_feedback: recentFeedbackSchema.nullable().optional(),
  has_project_links: z.boolean().optional(),
  processing_failed: z.boolean().optional(),
});

export const feedPaperSchema = paperSchema.extend({
  state: lifecycleStateSchema,
  state_before_trash: stateBeforeTrashSchema,
  starred: z.boolean(),
  rating: z.number().nullable(),
  summary_brief: z.string().nullable().optional(),
  tldr: z.string().nullable().optional(),
  confidence: confidenceSchema.nullable().optional(),
  priority_level: z.string().nullable().optional(),
  has_chunks: z.boolean().optional(),
  has_summary: z.boolean().optional(),
  recommendation_score: z.number().nullable().optional(),
  recommendation_reason: z.string().nullable().optional(),
  recommendation_modes: z.array(z.string()).nullable().optional(),
  note_match_count: z.number().optional(),
  note_snippet: z.string().nullable().optional(),
});

export const feedResponseSchema = z.looseObject({
  papers: z.array(feedPaperSchema),
  total: z.number(),
  search_mode: z.string().optional(),
});

export const paperBriefSchema = z.looseObject({
  id: z.number(),
  title: z.string(),
  source_type: z.string().nullable().optional(),
  published_date: z.string().nullable().optional(),
});

const extractedFieldSchema = z.looseObject({
  value: jsonValueSchema,
  quote: z.string().nullable(),
  verified: z.boolean(),
  confidence: z.number(),
  chunk_id: z.number().nullable(),
  page_number: z.number().nullable(),
});

export const extractionTableRowSchema = z.looseObject({
  paper_id: z.number(),
  paper_title: z.string(),
  extractions: z.record(z.string(), extractedFieldSchema),
});

export const batchExtractionAcceptedSchema = z.looseObject({
  job_id: z.string(),
  total: z.number(),
});

const searchPreviewLibraryMatchSchema = z.looseObject({
  paper_id: z.number(),
  has_project_links: z.boolean(),
  zotero_item_key: z.string().nullable(),
});

export const searchPreviewResultSchema = z.looseObject({
  external_id: z.string(),
  source_type: sourceTypeSchema,
  title: z.string(),
  authors: z.array(z.string()),
  abstract: z.string().nullable(),
  published_date: z.string().nullable(),
  url: z.string(),
  pdf_url: z.string().nullable(),
  citation_count: z.number(),
  metadata: jsonObjectSchema,
  library_match: searchPreviewLibraryMatchSchema.nullable(),
});

const searchPreviewSourceErrorSchema = z.looseObject({
  kind: z.enum(['rate_limit', 'api_error', 'unavailable']),
  message: z.string(),
  status_code: z.number().nullable(),
  retry_after_s: z.number().nullable(),
  settings_hint: z.string().nullable(),
});

export const searchPreviewResponseSchema = z.looseObject({
  results: z.array(searchPreviewResultSchema),
  total: z.number(),
  per_source_counts: z.record(z.string(), z.number()),
  degraded_sources: z.array(z.string()),
  source_errors: z.record(z.string(), searchPreviewSourceErrorSchema),
});

export const lifecycleActionResponseSchema = z.looseObject({
  status: z.literal('ok'),
  paper_id: z.number(),
});
export const hardDeleteResponseSchema = z.looseObject({ deleted: z.number() });
export const bulkActionResponseSchema = z.looseObject({
  succeeded: z.array(z.number()),
  failed: z.array(z.looseObject({ paper_id: z.number(), error: z.string() })),
});

export const feedCountsSchema = z.looseObject({
  inbox: z.number(),
  library: z.number(),
  reading_list: z.number(),
  reading: z.number(),
  done: z.number(),
  starred: z.number(),
  trash: z.number(),
  active: z.number(),
  kept: z.number(),
  all_non_trash: z.number(),
  by_source: z.record(z.string(), z.number()),
  by_topic: z.array(z.looseObject({ topic_id: z.number(), name: z.string(), count: z.number() })),
  untagged: z.number(),
});

export const feedbackResponseSchema = z.looseObject({
  paper_id: z.number(),
  signal: z.enum(['positive', 'negative']),
  source: feedbackSourceSchema,
  created_at: z.string(),
});
const feedbackListItemSchema = z.looseObject({
  paper_id: z.number(),
  title: z.string(),
  signal: z.enum(['positive', 'negative']),
  source: feedbackSourceSchema,
  reason: z.string().nullable(),
  topic_id: z.number().nullable(),
  topic_name: z.string().nullable(),
  created_at: z.string(),
});
export const feedbackListResponseSchema = z.looseObject({
  items: z.array(feedbackListItemSchema),
  total: z.number(),
});
export const deleteFeedbackResponseSchema = z.looseObject({
  deleted: z.number(),
  topic_id: z.number(),
});

export const discoveryResultSchema = z.looseObject({
  paper_id: z.number(),
  title: z.string(),
  authors: z.array(z.string()),
  matching_snippet: z.string(),
  similarity_score: z.number(),
  url: z.string(),
});

export const queuedJobSchema = z.looseObject({
  job_id: z.string(),
  status: z.literal('queued'),
  reason: z.null().optional(),
});
export const processLibraryResponseSchema = z.discriminatedUnion('status', [
  queuedJobSchema,
  z.looseObject({
    job_id: z.null(),
    status: z.literal('skipped'),
    reason: z.string(),
  }),
]);
export const batchProcessResponseSchema = z.looseObject({
  queued: z.number(),
  total_unprocessed: z.number(),
  skipped_missing_pdf: z.number(),
  job_id: z.string().nullable(),
});
export const batchSummarizeResponseSchema = z.looseObject({
  total_unsummarized: z.number(),
  job_id: z.string().nullable(),
});

export const noteSchema = z.looseObject({
  id: z.number(),
  paper_id: z.number(),
  user_note: z.string(),
  highlight_text: z.string().nullable(),
  page_number: z.number().nullable(),
  source: z.enum(['user', 'zotero']),
  zotero_annotation_key: z.string().nullable(),
  verification_status: z.enum(['unverified', 'verified', 'failed']),
  verified_quote: z.string().nullable(),
  verified_page_number: z.number().nullable(),
  promoted_at: z.string().nullable(),
  stale: z.boolean(),
  created_at: z.string(),
});

export const missingFoundationalPaperSchema = z.looseObject({
  paper_id: z.number(),
  title: z.string(),
  authors: z.array(z.string()),
  year: z.number().nullable(),
  citation_count: z.number(),
  cited_by_library_count: z.number(),
  url: z.string().nullable(),
  pdf_available: z.boolean(),
});
export const fetchAndProcessFoundationalSchema = z.looseObject({
  paper_id: z.number(),
  status: z.enum(['queued', 'no_pdf']),
  job_id: z.string().nullable(),
  message: z.string().nullable(),
});

export const citationRelationSchema = z.looseObject({
  source_paper_id: z.number(),
  cited_paper_id: z.number(),
  citation_context: z.string().nullable(),
  is_influential: z.boolean().nullable(),
  intent: z.array(z.string()),
});
const graphNodeSchema = z.looseObject({
  id: z.number(),
  title: z.string(),
  citation_count: z.number(),
  published_date: z.string().nullable(),
  is_stub: z.boolean(),
  display_size: z.number().optional(),
});
const graphEdgeSchema = z.looseObject({
  source: z.number(),
  target: z.number(),
  is_influential: z.boolean().nullable(),
  context: z.string().nullable(),
});
export const citationGraphSchema = z.looseObject({
  nodes: z.array(graphNodeSchema),
  edges: z.array(graphEdgeSchema),
});
export const citationFetchResponseSchema = z.looseObject({
  citations_added: z.number(),
  references_added: z.number(),
  stubs_created: z.number(),
});

export const entitySchema = z.looseObject({
  id: z.number(),
  name: z.string(),
  canonical_name: z.string(),
  entity_type: z.string(),
  description: z.string().nullable(),
  metadata: jsonObjectSchema,
  paper_count: z.number(),
  created_at: z.string().nullable(),
  display_size: z.number().optional(),
});
const relationshipSchema = z.looseObject({
  id: z.number(),
  source_entity_id: z.number(),
  target_entity_id: z.number(),
  relationship_type: z.string(),
  paper_id: z.number().nullable(),
  page_number: z.number().nullable(),
  evidence_quote: z.string().nullable(),
  confidence: z.number(),
  created_at: z.string().nullable(),
});
export const knowledgeGraphSchema = z.looseObject({
  entities: z.array(entitySchema),
  relationships: z.array(relationshipSchema),
  entity_type_counts: z.record(z.string(), z.number()).optional(),
});
export const kgQueryResponseSchema = z.looseObject({
  results: z.array(jsonObjectSchema),
  query: z.string(),
});
export const entityExtractionResponseSchema = z.looseObject({
  entities_added: z.number(),
  relationships_added: z.number(),
  entities_merged: z.number(),
  dropped_relationships: z.number().optional(),
  saved_by_full_text_verify: z.number().optional(),
});
export const batchEntityExtractionResponseSchema = z.looseObject({
  extracted: z.number(),
  failed: z.number(),
  total: z.number(),
});
